### PR TITLE
Add base logic

### PR DESCRIPTION
**Pull request by niusmallnan**
https://github.com/rancher/rancher/issues/9153

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/Sirupsen/logrus"
+)
+
+// APIServer structure is used to the store backend information
+type APIServer struct {
```
Comment (leodotcloud) [inline api.go]: Let's just get rid of this whole API server for now. We can revisit this later.
Two things I want to achieve later: 
- Improve upon https://github.com/leodotcloud/log to have an inbuilt server which listens on some configurable port, but listens only on localhost.
- Have a generic healthcheck framework and server that we can integrate in all of our microservices.

File: hostnat/watcher.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	w := &watcher{
+		c:         c,
+		ipsetName: setting.DefaultDisableHostNATIPset,
+	}
+	go c.OnChange(5, w.onChangeNoError)
+	return nil
+}
+
+type watcher struct {
+	mu        sync.Mutex
```
Comment (leodotcloud) [inline hostnat/watcher.go]: Do we really need this? We are not sharing this between different threads.

File: hostnat/watcher.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	allHosts, err := w.c.GetHosts()
+	if err != nil {
+		return err
+	}
+
+	return w.refreshIPSet(selfHost, allHosts)
+}
+
+func (w *watcher) refreshIPSet(selfHost metadata.Host, allHosts []metadata.Host) error {
+	ipsetPath, err := exec.LookPath("ipset")
```
Comment (leodotcloud) [inline hostnat/watcher.go]: We can have the assumption that ipset binary will be present, no need to check every time.

File: hostnat/watcher.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+		out, err = exec.Command(ipsetPath, "del", w.ipsetName, e, "-exist").CombinedOutput()
+		if err != nil {
+			optErr = errors.Wrap(optErr, err.Error())
+			continue
+		}
+	}
+	return optErr
+}
+
+func (w *watcher) diffIPSetEntries(ipsetPath string, selfHost metadata.Host, allHosts []metadata.Host) (toAddEntries []string, toDelEntries []string, err error) {
```
Comment (leodotcloud) [inline hostnat/watcher.go]: Inputs: current, desired
Outputs: toAdd, toDel

Passing in ipsetPath is not making this function modular.

File: hostnat/watcher.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	return toAddEntries, toDelEntries, nil
+}
+
+func (w *watcher) getCurrentIPSetEntries(ipsetPath string) (map[string]bool, error) {
+	currentEntries := map[string]bool{}
+	out, err := exec.Command(ipsetPath, "list", w.ipsetName).CombinedOutput()
+	if err != nil {
+		return currentEntries, errors.Wrapf(err, "Failed to list ipset %s: %s", w.ipsetName, out)
+	}
+	r := regexp.MustCompile("(?m)^(.*\n)*Members:\n")
```
Comment (leodotcloud) [inline hostnat/watcher.go]: looks like we can get the output of the list command in xml. `ipset list -o xml`
I guess parsing an XML output is more efficient than regex? May be you can write two functions, one using regex and another using xml parser and run them in *_test.go with a large ipset and check the performance?

File: hostnat/watcher.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		return errors.Wrapf(err, "Failed to create ipset: %s", out)
+	}
+
+	toAddEntries, toDelEntries, err := w.diffIPSetEntries(ipsetPath, selfHost, allHosts)
+	if err != nil {
+		return err
+	}
+	var optErr error
+	for _, e := range toAddEntries {
```
Comment (leodotcloud) [inline hostnat/watcher.go]: ipset supports swapping, not sure if it's very useful here. That's what we do in network-policy. If it helps you simply the code here, you can use it. Totally up to you.

File: main.go [hunk 7]
```diff
@@ -13,11 +19,70 @@ func main() {
 	app := cli.NewApp()
 	app.Name = "per-host-subnet"
 	app.Version = VERSION
-	app.Usage = "You need help!"
-	app.Action = func(c *cli.Context) error {
-		logrus.Info("I'm a turkey")
-		return nil
+	app.Usage = "support per-host-subnet networking"
+	app.Flags = []cli.Flag{
+		cli.StringFlag{
```
Comment (leodotcloud) [inline main.go]: As mentioned above, let's get rid of this for now.

File: main.go [hunk 8]
```diff
@@ -13,11 +19,70 @@ func main() {
...
+			return err
+		}
+	}
+
+	err = hostnat.Watch(m)
+	if err != nil {
+		return err
+	}
+
+	listenPort := ctx.String("listen")
```
Comment (leodotcloud) [inline main.go]: See above: remove.

File: package/Dockerfile [hunk 9]
```diff
@@ -1,3 +1,7 @@
-FROM ubuntu:16.04
+FROM rancher/agent-base:v0.3.0
```
Comment (leodotcloud) [inline package/Dockerfile]: We don't need `Dockerfile` for this. Can you please package it as a binary like ipsec/vxlan and make a separate PR to include this in `rancher/net`?

File: routeupdate/hostgw/hostgw.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+const (
+	ProviderName = "hostgw"
+
+	changeCheckInterval = 5
+	perHostSubnetLabel  = "io.rancher.network.per_host_subnet.subnet"
+)
+
+type HostGw struct {
+	mu sync.Mutex
```
Comment (leodotcloud) [inline routeupdate/hostgw/hostgw.go]: is the mutex needed?

File: routeupdate/hostgw/hostgw.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+	m  metadata.Client
+}
+
+func NewInst(m metadata.Client) (*HostGw, error) {
```
Comment (leodotcloud) [inline routeupdate/hostgw/hostgw.go]: `func NewRouteUpdater` or just `func New`?

**Review by leodotcloud — CHANGES_REQUESTED** [review]

File: package/Dockerfile (hunk 9 above)
Comment (niusmallnan) [reply package/Dockerfile]: Ok, I made a PR https://github.com/rancher/rancher-net/pull/78 .
Need merge `per-host-subnet` before you can review `rancher-net` PR.

**Review by leodotcloud — APPROVED** [review]
LGTM

**Pull request merged**